libraries/AP_HAL_ChibiOS/hwdef/common/hrt.c

```diff
@@ -68,48 +68,81 @@ static uint32_t system_time_u32_us(void)
 #error "unsupported timer resolution"
 #endif
 
-// offset for micros64()
-static uint64_t timer_base_us64;
-
 static uint32_t get_systime_us32(void)
 {
-    static uint32_t last_us32;
     uint32_t now = system_time_u32_us();
 #ifdef AP_BOARD_START_TIME
     now += AP_BOARD_START_TIME;
 #endif
-    if (now < last_us32) {
-        const uint64_t dt_us = 0x100000000ULL;
-        timer_base_us64 += dt_us;
-    }
-    last_us32 = now;
     return now;
 }
 
 /*
-  for the exposed functions we use chSysGetStatusAndLockX() to prevent
-  an interrupt changing the globals while allowing this call from any
-  context
+  for the exposed functions we use chVTGetTimeStampI which handles
+  wrap and directly gives a uint64_t (aka systimestamp_t)
 */
 
+uint64_t hrt_micros64I()
+{
+#ifdef AP_BOARD_START_TIME
+    return chVTGetTimeStampI() + AP_BOARD_START_TIME;
+#endif
+    return chVTGetTimeStampI();
+}
+
+static inline bool is_locked(void) {
+    return !port_irq_enabled(port_get_irq_status());
+}
+
 uint64_t hrt_micros64()
 {
-    syssts_t sts = chSysGetStatusAndLockX();
-    uint32_t now = get_systime_us32();
-    uint64_t ret = timer_base_us64 + now;
-    chSysRestoreStatusX(sts);
-    return ret;
+    if (is_locked()) {
+        return chVTGetTimeStampI();
+    } else if (port_is_isr_context()) {
+        uint64_t ret;
+        chSysLockFromISR();
+        ret = chVTGetTimeStampI();
+        chSysUnlockFromISR();
+        return ret;
+    } else {
+        uint64_t ret;
+        chSysLock();
+        ret = chVTGetTimeStampI();
+        chSysUnlock();
+        return ret;
+    }
 }
 
 uint32_t hrt_micros32()
 {
-    syssts_t sts = chSysGetStatusAndLockX();
-    uint32_t ret = get_systime_us32();
-    chSysRestoreStatusX(sts);
-    return ret;
+#if CH_CFG_ST_RESOLUTION == 16
```

Comment:
I don't understand why this is only for 16-bit timers - a comment would be nice

Reply:
comment added